test/junit/scala/concurrent/FutureTest.scala

```diff
@@ -6,8 +6,47 @@ import org.junit.Test
 
 import scala.tools.testkit.AssertUtil._
 import scala.util.Try
+import duration.Duration.Inf
 
 class FutureTest {
+  @Test
+  def testZipWithFailFastBothWays(): Unit = {
+    import ExecutionContext.Implicits.global
+
+    val start = System.currentTimeMillis()
+    val f1 = Future{Thread.sleep(1000); throw new Exception("Boom Late"); 123}
+    val f2 = Future{Thread.sleep(200); throw new Exception("Boom Early"); 456}
+
+    val scala.util.Failure(fa) = Try(Await.result(f1.zip(f2), Inf))
+    val scala.util.Failure(fb) = Try(Await.result(f2.zip(f1), Inf))
+
+    val scala.util.Failure(fc) = Try(Await.result(f1.zipWith(f2)((_, _)), Inf))
+    val scala.util.Failure(fd) = Try(Await.result(f2.zipWith(f1)((_, _)), Inf))
+
+    val scala.util.Failure(fe) = Try(Await.result(Future.sequence(Seq(f1, f2)), Inf))
+    val scala.util.Failure(ff) = Try(Await.result(Future.sequence(Seq(f2, f1)), Inf))
+
+    val scala.util.Failure(fg) = Try(Await.result(Future.traverse(Seq(0, 1))(Seq(f1, f2)(_)), Inf))
+    val scala.util.Failure(fh) = Try(Await.result(Future.traverse(Seq(0, 1))(Seq(f1, f2)(_)), Inf))
+
+    val end = System.currentTimeMillis()
+    // Make sure the early failure is always reported, regardless of whether it's on
+    // the left or right of the zip/zipWith/sequence/traverse
+    assert(fa.getMessage == "Boom Early")
+    assert(fb.getMessage == "Boom Early")
+    assert(fc.getMessage == "Boom Early")
+    assert(fd.getMessage == "Boom Early")
+    assert(fe.getMessage == "Boom Early")
+    assert(ff.getMessage == "Boom Early")
+    assert(fg.getMessage == "Boom Early")
+    assert(fh.getMessage == "Boom Early")
+
+    // Make sure that it fails early, after the earlier failure occurs, and does not wait
+    // for the later failure regardless of which one is on the left and which is on the right
+    assert(end - start > 150)
+    assert(end - start < 950)
```

Comment:
I think that there's a risk this will error out on slow machines.
Also, do not use `currentTimeMillis` for checking elapsed time, use `nanoTime` as `currentTimeMillis` has limited accuracy on many OS:es